src/Microsoft.AspNetCore.Server.IISIntegration/Server/IO/AsyncIOEngine.cs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration
+{
+    internal partial class AsyncIOEngine : IAsyncIOEngine
+    {
+        private readonly IntPtr _handler;
+
+        private AsyncIOOperation _nextOperation;
+        private AsyncIOOperation _runningOperation;
+
+        private AsyncReadOperation _cachedAsyncReadOperation;
+        private AsyncWriteOperation _cachedAsyncWriteOperation;
+        private AsyncFlushOperation _cachedAsyncFlushOperation;
+
+        public AsyncIOEngine(IntPtr handler)
+        {
+            _handler = handler;
+        }
+
+        public ValueTask<int> ReadAsync(Memory<byte> memory)
+        {
+            var read = GetReadOperation();
+            read.Initialize(_handler, memory);
+            Run(read);
+            return new ValueTask<int>(read, 0);
+        }
+
+        public ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+        {
+            var write = GetWriteOperation();
+            write.Initialize(_handler, data);
+            Run(write);
+            CancelPendingRead();
+            return new ValueTask<int>(write, 0);
+        }
+
+        private void Run(AsyncIOOperation ioOperation)
+        {
+            lock (this)
+            {
+                if (_runningOperation != null)
+                {
+                    if (_nextOperation == null)
+                    {
+                        _nextOperation = ioOperation;
```

Comment:
This appears to allow for two reads or two writes?

Reply:
It allows for one operation to be running and another queued. This is required so we can wait until current read cancellation notification would arrive and then pick up the next write.